Simplenote/src/main/java/com/automattic/simplenote/utils/TagUtils.java

```diff
@@ -165,4 +169,42 @@ private static String replaceEncoded(String encoded) {
             .replace(".", "%2E")
             .replace("_", "%5F");
     }
+
+    /**
+     * Rename a tag in a note, given the name of the tag to be renamed and the new tag name.
+     * @param tagOld The tag to be renamed.
+     * @param tagNew The tag to replace the old tag.
+     * @param note The note object to be modified.
+     */
+    public static void renameTagInNote(Note note, String tagOld, String tagNew) {
+        List<String> tags = note.getTags();
+
+        List<String> tagHashList = new ArrayList<>();
+        for (String tag : tags) {
+            tagHashList.add(TagUtils.hashTag(tag));
+        }
+
+        String tagOldHash = TagUtils.hashTag(tagOld);
+        String tagNewHash = TagUtils.hashTag(tagNew);
+        boolean isOldHashEqualNew = tagOldHash.equals(tagNewHash);
+
+        ListIterator<String> iterator = tags.listIterator();
+        while (iterator.hasNext()) {
+            String tag = iterator.next();
+
+            if (tag.equals(tagOld)) {
+                // Remove old tag
+                iterator.remove();
+
+                if (isOldHashEqualNew || !tagHashList.contains(tagNewHash)) {
+                    // Add new tag
+                    iterator.add(tagNew);
+                }
+            }
+
+        }
+
+        note.setTags(tags);
+    }
```

Comment:
I'm still a bit confused with this function, specifically the multiple iterations, which I don't think are necessary. here's another formulation that I think is supposed to do the same thing, though it covers one more case, which is where we already had a tag in a note that matches the new name of the renamed tag. this situation represents certain data corruption, but it definitely exists in practice.

```erlang
public static void renameTagInNote(Note note, String tagOld, String tagNew) {
    List<String> newTags = new ArrayList<>();
    String oldHash = hashTag(tagOld);
    String newHash = hashTag(tagNew);

    for (String tag : note.getTags()) {
        String hash = hashTag(tag);

        if (hash.equals(newHash)) {
            continue;
        }

        if (hash.equals(oldHash)) {
            newTags.add(tagNew);
            continue;
        }

        newTags.add(tag);
    }

    note.setTags(newTags);
}
```

Reply:
I think the suggested implementation doesn't support renaming a tag to a lexical variation. But the old implementation does.

We'd also have to support the following two scenarios since the old implementation also supports those.

| Scenario                                | Tags              | Rename       | Expected Result    |
| -----------------------                  | ----------------- | ---------------- | ----------    |
| Rename to lexical                 | one               | one  -->  ONE    | ONE        |
| Rename to existing in a lexical form   | one two         | one  -->  TWO | two          |


    @Test
    public void renameToLexical() {
        Note note = new Note("key");
        note.setTags(tagList("one"));

        TagUtils.renameTagInNote(note, "one", "ONE");
        assertThat(note.getTags(), is(tagList("ONE")));
    }

    @Test
    public void renameToExistingLexical() {
        Note note = new Note("key");
        note.setTags(tagList("one", "two"));

        TagUtils.renameTagInNote(note, "one", "TWO");
        assertThat(note.getTags(), is(tagList("two")));
    }

I actually just refactor and restructure the logic that is used to rename a tag in a note. `TagUtilsTest#testRenameTagInNote` test is being added to make sure that the final outcome is as same as the old code. The reason I used an iterator is to maintain the order of the tags.

For instance previously if we rename a tag, the renamed tag is being appended to the end of the tag string, like: 
 
| Tags              | Rename       |  Result    |
 | ----------------- | ---------------- | ----------    |
| One two three | One  -->  one | two three one      |

Please have a look at the unit test `TagUtilsTest#testRenameTagInNote` and let me know if `TagUtils#renameTagInNote` method can be further simplified.